service-base/src/test/java/org/eclipse/hono/service/management/device/AutoProvisioningEnabledDeviceBackendTest.java

```diff
@@ -92,4 +101,45 @@ void provisionDevice(final VertxTestContext ctx) {
         }));
     }
 
+    /**
+     * Verifies that if the creation of credentials fails during the the provisioning, the previously created device
+     * will be removed.
+     * 
+     * @param ctx The vert.x test context.
+     */
+    @Test
+    public void testProvisionDeviceRemovesDeviceIfCredentialsCreationFails(final VertxTestContext ctx) {
+
+        // GIVEN an AutoProvisioningEnabledDeviceBackend instance where the creation of credentials fails
+        final AutoProvisioningEnabledDeviceBackend underTest = mock(AutoProvisioningEnabledDeviceBackend.class);
+        when(underTest.provisionDevice(anyString(), any(), any())).thenCallRealMethod();
+
+        doAnswer(invocation -> {
+            final Promise<OperationResult<Id>> promise = invocation.getArgument(4);
+            promise.complete(OperationResult.ok(201, Id.of(DEVICE_ID), Optional.empty(), Optional.empty()));
+            return null;
+        }).when(underTest).createDevice(any(), any(), any(), any(), any(Promise.class));
+
+        doAnswer(invocation -> {
+            final Promise<Result<Void>> promise = invocation.getArgument(4);
+            promise.complete(Result.from(204));
+            return null;
+        }).when(underTest).deleteDevice(any(), any(), any(), any(), any(Promise.class));
+
+        doAnswer(invocation -> {
+            final Promise<OperationResult<Void>> promise = invocation.getArgument(5);
+            promise.complete(OperationResult.empty(403)); // creation of credentials fails
+            return null;
+        }).when(underTest).set(any(), any(), any(), any(), any(), any(Promise.class));
+
+        // WHEN provisioning a device from a certificate
+        final Future<OperationResult<String>> result = underTest.provisionDevice(TENANT_ID, cert, NoopSpan.INSTANCE);
+
+        // THEN the device is deleted
+        result.setHandler(ctx.succeeding(ok -> {
+            verify(underTest).deleteDevice(eq(TENANT_ID), eq(DEVICE_ID), any(), any(), any(Promise.class));
```

Comment:
this needs to go into a
```
ctx.verify(() -> {
  ...
});
```
block in order for the ctx to fail in case of an assertion error.